### ISSUE TITLE
Graphql transformer does not generate query filters for nested type objects

### ISSUE THREAD

**Issue by vkumarvs** [aws-amplify/amplify-cli#2055]
>**Note**: If your issue/bug is regarding the AWS Amplify Console service, please log it in the 
[Amplify Console GitHub Issue Tracker](https://github.com/aws-amplify/amplify-console/issues)

**Describe the bug**
I am trying to test appsync auto generated resolvers and queries for below schema. The problem I am facing, is that the query filter doesn't include filter for nested type ("enterprise" , it entered into dynamo as a Map attribute ). 

type TestAuth @model
    @key(fields: ["pk", "sk"])
{
######################################
    pk                 :ID!
    sk                 :ID!
    enterprise         :EnterpriseData
}

type EnterpriseData {
    tenantId           :String
    tenantName         :String
}
Any suggestion or workaround would be greatly helpful. I don't want to write customise filters.

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by kaustavghosh06** [comment]
@vkumarvs You can model your transformer schema like the following:

```graphql
type TestAuth @model
@key(fields: ["pk", "sk"])
{
######################################
pk :ID!
sk :ID!
enterprise :EnterpriseData @connection
}

type EnterpriseData @model {
tenantId :String
tenantName :String
}
```

And then from your client you can call a query  in this format:

```graphql
Query testAuthENt {
  getTestAuth(filter: ...) {
    enterprise(filter: ...) {
      
    }
  }
}
```

**Comment by vkumarvs** [comment]
> @vkumarvs You can model your transformer schema like the following:
> 
> ```graphql
> type TestAuth @model
> @key(fields: ["pk", "sk"])
> {
> ######################################
> pk :ID!
> sk :ID!
> enterprise :EnterpriseData @connection
> }
> 
> type EnterpriseData @model {
> tenantId :String
> tenantName :String
> }
> ```
> 
> And then from your client you can call a query in this format:
> 
> ```graphql
> Query testAuthENt {
>   getTestAuth(filter: ...) {
>     enterprise(filter: ...) {
>       
>     }
>   }
> }
> ```

Thank you @kaustavghosh06 for reply but I don't want to use two tables for this. My data will remain in single table. 
More importantly , it seems amplify doesn't autogenerate nested type filtering, nested types makes my single table schema look clean but I can't query those fields. It sound to me a feature request.

**Comment by dahersoftware** [comment]
@kaustavghosh06  any update on this?

**Comment by kaustavghosh06** [comment]
@vkumarvs That's correct. DynamoDB doesn't provide filtering capabilities for complex JSON objects. For advanced filtering you could either use Elasticsearch along with your DDB using @searchable dirctive or use two models.

**Comment by dahersoftware** [comment]
@kaustavghosh06 I will explore @searchable directive but graphql support for nested objects would have been amazing. I just need to fetch all my complex objects rows that are keyed and include the rows that matches my nested attributes field. 
Creating multiple tables to satisfy only this requirement sound little harsh to me.  

My schema is something like below.
type Factory
  @model
  @key(fields: ["pk", "sk"])
  ) {
  pk: ID!
  sk: ID!
  dataType: DataType!
  ######################################
  workstation: WorkStationData #input types nested objects
  region: RegionData #input type nested object
}
I would have listed all the nested object based on my PK and sk and then further filter based on nested object (workstations) attribute.)

**Comment by vkumarvs** [comment]
type Factory
  @model
  @searchable
  @key(fields: ["pk", "sk"])
  ) {
  pk: ID!
  sk: ID!
  dataType: DataType!
  ######################################
  workstation: WorkStationData #input types nested objects
  region: RegionData #input type nested object
}

I used searchable directive and the filtering code generated by Tranformer doesn't include workstation object.

export type SearchableFactoryFilterInput = {
  pk?: SearchableIDFilterInput | null,
  sk?: SearchableIDFilterInput | null,
  and?: Array< SearchableFactoryFilterInput | null > | null,
  or?: Array< SearchableFactoryFilterInput | null > | null,
  not?: SearchableFactoryFilterInput | null,
};      

It only contains pk and sk so I do not have choice to get data based on nested filtering. Any suggestion?

**Comment by evrardts** [comment]
same problem here, would be great to have @searchable working with nested fields

**Comment by ashish369in** [comment]
Same request here , it would be great to have @searchable working with nested fields.

**Comment by IsaacTrevino** [comment]
+1

**Comment by kaxax** [comment]
+1

**Comment by Jupdi** [comment]
+1 
Important feature! We need to setup our own search instance to get this working

**Comment by josefaidt** [comment]
per @undefobj we should also think about this for the `_version` and `_deleted` filters when using GraphQL API + DataStore

**Comment by Lacie** [comment]
+1

**Comment by banut1** [comment]
+1
